### ISSUE TITLE
Killing remote disk due to way too high concurrency

### ISSUE THREAD

**Issue by zcalusic** [restic/restic#668]
This is manually built latest development version of restic (gb build).

I noticed that, when using restic-server backend, restic fires up to 40 concurrent connections during check command. Then all these connections translate to read I/O requests on the remote machine. This is way way too high for a single rotational hard drive on the destination. Heck, even SSD there would have trouble to cope with this kind of load.

The result of this is I/O starvation, load average on the machine skyrockets, and various other stuff that runs on that machine starts failing, not being able to reach disk for a long time, fighting through all this outstanding I/O requests. Even restic is definitely running much slower than it could, because so many synchronous I/O request at the same time basically push the I/O pattern to random, and disk throughput drastically falls.

I see that REST backend has connLimit = 10, which is also a tad high, but even that limit is not obeyed, so there's some bug in that whole mechanism of dispatching commands to restic-server.

Could we make this parallelism tunable? SSD's can stand a bit of parallelism, but rotational disks are best if left with no more than 1 or max 2 outstanding request at a time, anything more and throughput will fall. I guess I could add this tunable in the restic-server itself, but somehow I think it would be better if backup client would behave better, to start with.

**Comment by fd0** [comment]
I agree and I'm currently rewriting the archiver code. The IO concurrency will be reduced and I'll probably also make this tuneable.

**Comment by zcalusic** [comment]
Ha, I found where the magical number 40 comes from: https://github.com/restic/restic/blob/master/src/restic/checker/checker.go#L53

Funny thing, now with perfectly cached files on both sides due to b56bde3f618789030f5712cf9fc3a962e9ee9d83 commit, I can't even reproduce what I saw the other day. :) Wanted to test the difference if I replace 40 with 10... But, with Linux page cache in effect again restic check is now freaking fast, so I can't catch the high concurrency atm, will need much bigger repo for that to start happening again. :smile:

I guess I'll let you sort this issue out.

Having a user controlled tunable is an excellent idea. :+1: One HDD != 6 disk RAID != SSD/NVMe, everybody's requirements will be different, and tunable will help to adapt to various scenarios easily.

**Comment by fd0** [comment]
On Linux, it is possible to ask the kernel whether or not a device is a harddisk:
```
$ cat /sys/block/sda/queue/rotational
0
```
See http://unix.stackexchange.com/a/65602

**Comment by wscott** [comment]
It is, but it is tricky.  Here is our attempt to answer that question:
https://github.com/bitkeeper-scm/bitkeeper/blob/master/src/port/networkfs.c

Sometimes it doesn't know.

On Fri, Nov 11, 2016 at 5:25 AM, Alexander Neumann <notifications@github.com

**Comment by zcalusic** [comment]
Once again, I don't think backup program should be so clever and peek into hardware capabilities. If you go that route, i.e. that restic gets to decide about concurrency, not operator, than IMHO you need to answer ALL these questions:

- how many disks are behind HP hardware raid, 5 or 50? which RAID level 5 or 10 (5 has much worse write performance)
- the above for LVM, software RAID, LSI, Adaptec... and BTW, how big is the cache, 2GB or 16GB?
- are the disks 5400rpm SATA or 15k SAS/FC (a BIG difference in performance)?
- what is the workload of the storage, is it mostly idle, or under heavy load?

and finally, my favorite :smile:, it needs to **read my mind**: 

- it needs to understand that on server A I want it to go as fast as it can, because backup is of the outmost importance there
- it needs to be invisible on server B, there it is only to offer disaster recovery, but it should under no circumstances interfere with primary workload

tl;dr IMHO, don't make backup too clever, just make it tunable, it's easier and more useful in practice

**Comment by fd0** [comment]
In general I agree with @zcalusic. I just stumbled upon the capability of the kernel that I didn't know about. I have no intentions to use it in restic.

As a general note, I'd like to repeat what I wrote in earlier issues (that I cannot seem to find at the moment, unfortunately): I have a vision for restic, that among others things it should be easy to use and work out of the box for most cases.

Regarding simplicity, in my opinion sometimes it's best to not give users too many options to choose from. Pick a sane default that works for most use cases, and offer options for other (important) use cases not covered by the default solution. Especially, this means that not everything will be configurable. This is different from what many other programs (especially CLI programs) do, and I know that sometimes this is not the most popular way to go.

So for the number of IO operations I think a parameter makes sense, with a sane default of 1-2 concurrent operations. I'm currently reworking the archiver code which is used to create a backup (at the moment it reads up to 10 files in parallell), we'll see how that goes.

**Comment by rawtaz** [comment]
@zcalusic Is this issue still a thing/issue? The archiver was reworked in 0.9.0 so I'm wondering if this issue still applies. Otherwise we should close it.

**Comment by zcalusic** [comment]
You're right @rawtaz I had no issues on this front for a long time. For a while I ran a patched version which lowered then default concurrency of 40 to only 2, and later on I think I saw the default concurrency was changed to 5, at which point I again tested the official release, and it has been running happily ever since. :)
